hydra-node/src/Hydra/Network/Etcd.hs

```diff
@@ -0,0 +1,513 @@
+              -- TODO: DRY just cases
+              Just
+                HydraHandshakeRefused
+                  { remoteHost = Host "???" $ fromMaybe 0 $ parsePort port
+                  , ourVersion
+                  , theirVersions = KnownHydraVersions [theirVersion]
+                  }
+        Nothing ->
+          Just
+            HydraHandshakeRefused
+              { remoteHost = Host "???" $ fromMaybe 0 $ parsePort port
+              , ourVersion
+              , theirVersions = NoKnownHydraVersions
+              }
+    _ ->
+      Just
+        HydraHandshakeRefused
+          { remoteHost = Host "???" 0 -- TODO: use optional data type
+          , ourVersion
+          , theirVersions = NoKnownHydraVersions
+          }
+ where
+  parseVersion = fmap MkHydraVersionedProtocolNumber . readMaybe . toString
+
+  parsePort = readMaybe . toString
+
+-- | Broadcast messages from a queue to the etcd cluster.
+--
+-- TODO: retrying on failure even needed?
+-- Retries on failure to 'putMessage' in case we are on a minority cluster.
+broadcastMessages ::
+  (ToCBOR msg, Eq msg) =>
+  Tracer IO EtcdLog ->
+  Connection ->
+  HydraVersionedProtocolNumber ->
+  PortNumber ->
+  PersistentQueue IO msg ->
+  IO ()
+broadcastMessages tracer conn protocolVersion port queue =
+  withGrpcContext "broadcastMessages" . forever $ do
+    msg <- peekPersistentQueue queue
+    (putMessage conn protocolVersion port msg >> popPersistentQueue queue msg)
+      `catch` \case
+        GrpcException{grpcError = GrpcUnavailable, grpcErrorMessage} -> do
+          traceWith tracer $ BroadcastFailed{reason = fromMaybe "unknown" grpcErrorMessage}
+          threadDelay 1
+        e -> throwIO e
+
+-- | Broadcast a message to the etcd cluster.
+putMessage ::
+  ToCBOR msg =>
+  Connection ->
+  HydraVersionedProtocolNumber ->
+  PortNumber ->
+  msg ->
+  IO ()
+putMessage conn protocolVersion port msg =
+  void $ nonStreaming conn (rpc @(Protobuf KV "put")) req
+ where
+  req =
+    defMessage
+      & #key .~ key
+      & #value .~ serialize' msg
+
+  -- TODO: use one key again (after mapping version check)?
+  key = encodeUtf8 @Text $ "msg-" <> show (hydraVersionedProtocolNumber protocolVersion) <> "-" <> show port
+
+-- | Fetch and wait for messages from the etcd cluster.
+waitMessages ::
+  FromCBOR msg =>
+  Connection ->
+  HydraVersionedProtocolNumber ->
+  FilePath ->
+  NetworkCallback msg IO ->
+  IO ()
+waitMessages conn protocolVersion directory NetworkCallback{deliver, onConnectivity} = do
+  revision <- getLastKnownRevision directory
+  withGrpcContext "waitMessages" . forever $ do
+    biDiStreaming conn (rpc @(Protobuf Watch "watch")) $ \send recv -> do
+      -- NOTE: Request all keys starting with 'msg'. See also section KeyRanges
+      -- in https://etcd.io/docs/v3.5/learning/api/#key-value-api
+      let watchRequest =
+            defMessage
+              & #key .~ "msg"
+              & #rangeEnd .~ "msh" -- NOTE: g+1 to query prefixes
```

Comment:
Why is this needed?

Reply:
Which "this" do you mean? What is unclear?

The NOTE on this line says g+1 (and refers to 'h') to get the prefixes.
The NOTE on this variable assignment explains that we "request all keys starting with 'msg'" and gives a link with further details.